dist/systemd/system/greenboot-rollback.service

```diff
@@ -1,9 +1,10 @@
 
 [Unit]
 Description=Greenboot rollback
-Before=multi-user.target greenboot.service
+DefaultDependencies=no
+Before=multi-user.target ostree-finalize-staged.service greenboot.service
 Wants=local-fs.target
-After=local-fs.target systemd-update-done.service
```

Comment:
https://www.freedesktop.org/software/systemd/man/systemd-update-done.service.html

this is very much needed isn't it? otherwise `ConditionNeedsUpdate` won't work (as that relies on the mtime in these dirs too)

Reply:
As per systemd docs  any service that should run after upgrade needs to be ordered as `Before`  systemd-update-done.service  but the service  `Conflicts=shutdown.target` which rollback tries to achieve, this is probably the reason things get stuck during the rollback process.

Reply:
Tested it again with reintroducing this line, system-update-done gets completed on the 1st restart when the greenboot-health fails for the 1st time, which causes the rollback to miss its trigger for the subsequent restart.
So I think we cannot avoid the `DefaultDependencies=no` for this case and if we add `Before=system-update-done.done` we can opt for executing `/sbin/reboot` directly instead of `systemctl reboot` which I am not sure we want to do.

```
[admin@localhost ~]$ sudo systemctl status greenboot-rollback.service
○ greenboot-rollback.service - Greenboot rollback
     Loaded: loaded (/usr/lib/systemd/system/greenboot-rollback.service; enabled; preset: disabled)
     Active: inactive (dead)
  Condition: start condition failed at Thu 2023-05-04 05:33:27 EDT; 38s ago
             ├─ ConditionNeedsUpdate=|/etc was not met
             └─ ConditionNeedsUpdate=|/var was not met

May 04 05:33:27 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -0 -u greenboot-rollback.service
May 04 05:33:27 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -1 -u greenboot-rollback.service
May 04 05:31:47 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -2 -u greenboot-rollback.service
May 04 05:30:05 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -3 -u greenboot-rollback.service
May 04 05:28:48 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -4 -u greenboot-rollback.service
May 04 05:25:46 localhost systemd[1]: Greenboot rollback was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -0 -u greenboot
May 04 05:33:27 localhost systemd[1]: Starting greenboot Health Checks Runner...
May 04 05:33:27 localhost greenboot[797]:  INFO  greenboot > GreenbootConfig { max_reboot: 1 }
May 04 05:33:27 localhost greenboot[797]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/01_repository_dns_check.sh
May 04 05:33:27 localhost greenboot[797]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/02_watchdog.sh
May 04 05:33:27 localhost greenboot[797]:  INFO  greenboot > running wanted check /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh
May 04 05:33:27 localhost greenboot[797]:  WARN  greenboot > wanted script /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh failed!
May 04 05:33:27 localhost greenboot[797]:  INFO  greenboot > greenboot health-check passed.
May 04 05:33:28 localhost systemd[1]: Finished greenboot Health Checks Runner.
[admin@localhost ~]$ sudo journalctl -b -1 -u greenboot
May 04 05:31:47 localhost systemd[1]: Starting greenboot Health Checks Runner...
May 04 05:31:47 localhost greenboot[804]:  INFO  greenboot > GreenbootConfig { max_reboot: 1 }
May 04 05:31:47 localhost greenboot[804]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/01_repository_dns_check.sh
May 04 05:31:47 localhost greenboot[804]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/02_watchdog.sh
May 04 05:31:47 localhost greenboot[804]:  INFO  greenboot > running required check /etc/greenboot/check/required.d/10_failing_check.sh
May 04 05:31:47 localhost greenboot[804]:  ERROR greenboot > required script /etc/greenboot/check/required.d/10_failing_check.sh failed!
May 04 05:31:47 localhost greenboot[804]:  INFO  greenboot > running wanted check /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh
May 04 05:31:47 localhost greenboot[804]:  WARN  greenboot > wanted script /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh failed!
May 04 05:31:47 localhost greenboot[804]:  ERROR greenboot > Greenboot health-check failed!
May 04 05:31:47 localhost greenboot[804]:  ERROR greenboot > cannot set boot_counter as: boot_counter already set
May 04 05:31:47 localhost greenboot[804]: Error: health-check failed!
May 04 05:31:47 localhost systemd[1]: greenboot.service: Main process exited, code=exited, status=1/FAILURE
May 04 05:31:47 localhost systemd[1]: greenboot.service: Failed with result 'exit-code'.
May 04 05:31:47 localhost systemd[1]: Stopped greenboot Health Checks Runner.
[admin@localhost ~]$ sudo journalctl -b -2 -u greenboot
May 04 05:30:05 localhost systemd[1]: Starting greenboot Health Checks Runner...
May 04 05:30:05 localhost greenboot[809]:  INFO  greenboot > GreenbootConfig { max_reboot: 1 }
May 04 05:30:05 localhost greenboot[809]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/01_repository_dns_check.sh
May 04 05:30:05 localhost greenboot[809]:  INFO  greenboot > running required check /usr/lib/greenboot/check/required.d/02_watchdog.sh
May 04 05:30:05 localhost greenboot[809]:  INFO  greenboot > running required check /etc/greenboot/check/required.d/10_failing_check.sh
May 04 05:30:05 localhost greenboot[809]:  ERROR greenboot > required script /etc/greenboot/check/required.d/10_failing_check.sh failed!
May 04 05:30:05 localhost greenboot[809]:  INFO  greenboot > running wanted check /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh
May 04 05:30:05 localhost greenboot[809]:  WARN  greenboot > wanted script /usr/lib/greenboot/check/wanted.d/01_update_platforms_check.sh failed!
May 04 05:30:05 localhost greenboot[809]:  ERROR greenboot > Greenboot health-check failed!
May 04 05:30:05 localhost greenboot[809]: Error: health-check failed!
May 04 05:30:06 localhost systemd[1]: greenboot.service: Main process exited, code=exited, status=1/FAILURE
May 04 05:30:06 localhost systemd[1]: greenboot.service: Failed with result 'exit-code'.
May 04 05:30:06 localhost systemd[1]: Stopped greenboot Health Checks Runner.
[admin@localhost ~]$ sudo journalctl -b -0 -u systemd-update-done.service
May 04 05:33:27 localhost systemd[1]: Update is Completed was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -1 -u systemd-update-done.service
May 04 05:31:46 localhost systemd[1]: Update is Completed was skipped because no trigger condition checks were met.
[admin@localhost ~]$ sudo journalctl -b -2 -u systemd-update-done.service
May 04 05:30:05 localhost systemd[1]: Starting Update is Completed...
May 04 05:30:05 localhost systemd[1]: Finished Update is Completed.
May 04 05:31:36 localhost systemd[1]: systemd-update-done.service: Deactivated successfully.
May 04 05:31:36 localhost systemd[1]: Stopped Update is Completed.
```

Reply:
@runcom thanks for sparking the question about `ConditionNeedsUpdate`,  went through https://www.freedesktop.org/software/systemd/man/systemd.unit.html#ConditionNeedsUpdate=
it seemd like this is what we only need as this happens only when an offline upgrade happens update only and it doesn't hinder with the shutdown.target. updated the code and verified it.